storage_controller/src/compute_hook.rs

```diff
@@ -614,6 +645,36 @@ impl ComputeHook {
         self.notify_execute(maybe_send_result, tenant_shard_id, cancel)
             .await
     }
+
+    /// Reflect a detach for a particular shard in the compute hook state.
+    ///
+    /// The goal is to avoid sending compute notifications with stale information (i.e.
+    /// including detach pageservers).
+    #[tracing::instrument(skip_all, fields(tenant_id=%tenant_shard_id.tenant_id, shard_id=%tenant_shard_id.shard_slug()))]
+    pub(super) fn handle_detach(
+        &self,
+        tenant_shard_id: TenantShardId,
+        stripe_size: ShardStripeSize,
+    ) {
+        use std::collections::hash_map::Entry;
+
+        let mut state_locked = self.state.lock().unwrap();
+        match state_locked.entry(tenant_shard_id.tenant_id) {
+            Entry::Vacant(_) => {
+                tracing::warn!("Compute hook tenant not found for detach");
+            }
+            Entry::Occupied(mut e) => {
+                let sharded = e.get().is_sharded();
+                if !sharded {
+                    e.remove();
+                } else {
+                    e.get_mut().handle_detach(tenant_shard_id, stripe_size);
```

Comment:
I think we are now implementing support for detaching shard-by-shard but I am unsure why would we ever support that vs. detaching all shards.

Regardless, isn't here an unhandled case where the `e` becomes empty?

Reply:
> I think we are now implementing support for detaching shard-by-shard but I am unsure why would we ever support that vs. detaching all shards.

Detach of single shards is the primitive that enables detaching the entire tenant. It's already implemented and required for cutting over to the secondary location. Manual detach of a tenant is implemented by handling each shard in turn [here](https://github.com/neondatabase/neon/blob/9d1c6f23d312f43ecd0d7134d7566020626548b2/storage_controller/src/service.rs#L2932-L2950).

> Regardless, isn't here an unhandled case where the e becomes empty or are we supposed to hold on to a husk of sharded shardless ComputeHookTenant?

Indeed, good spot. I checked the update path before writing the code. When removing the last shard we can either leave the state as is (empty) or remove it from the map altogether. Both cases are supported on the update path [here](https://github.com/neondatabase/neon/blob/main/storage_controller/src/compute_hook.rs#L86-L102).